### PR TITLE
Report to list

### PR DESCRIPTION
**Pull request by saulshanabrook**
**Not ready for merge**

### CODE REVIEW

**Comment by grantjenks** [comment]
Is this intended to replace the report_to_list method in django-report-utils? What's happening with that module? There are a number of refactorings there which aren't present in this branch.

**Comment by saulshanabrook** [comment]
@grantjenks Yes that is the plan. @bufke was saying that `django-report-utils` seemed to be heavily linked to this package anyway, so some of the code could be simplified by moving it inside.

@bufke The plan would be then to bring any important changes from `django-report-utils` inside this repos implementation as well?

**Comment by grantjenks** [comment]
There's certainly some awkwardness in how the logic is spread-out.

I have one more test to add regarding group-by and aggregates. Let me get that in before this so we don't regress. I'll create a pull-request today.

**Comment by grantjenks** [comment]
I was wrong. All my tests are in. I don't have another to add.

**Comment by grantjenks** [comment]
Note my recent changes in #186 I'd like that feature / those tests to still work after the refactoring.

Also note that the implementation for group-by support in is report_to_list in report_utils. I don't think that's in this refactoring though it should be moved there. Probably get_query was a more appropriate place for that change.

**Comment by bufke** [comment]
@grantjenks not looking to break any features. Your unit tests are very helpful. The is not near ready to push yet. 

@saulshanabrook I did a new git rebase - can you confirm I did it correctly? git push -f scares me.

**Comment by saulshanabrook** [comment]
@bufke looks like it worked. i would squash all commits into one commit, or a couple of logical ones. Right now it seems a bit hard to follow:
screen shot 2015-07-12 at 12 51 21 pm

**Comment by bufke** [comment]
Ok all unit tests pass. Released to pypi as 3.2.0rc1.

@grantjenks if you have time could you install the RC and confirm it's not breaking things.

File: docker-compose.yml [hunk 1]
```diff
@@ -7,7 +7,7 @@ web:
   command: python manage.py runserver_plus 0.0.0.0:8000
   volumes:
     - .:/code
-    #- ../django-report-utils/report_utils:/usr/local/lib/python3.4/site-packages/report_utils
+    - ../django-report-utils/report_utils:/usr/local/lib/python3.4/site-packages/report_utils
```
Comment (saulshanabrook) [inline docker-compose.yml]: Did you mean to leave this un commented?

File: report_builder/models.py [hunk 2]
```diff
@@ -85,18 +88,18 @@ def save(self, *args, **kwargs):
...
-                queryset = queryset.annotate(Avg(display_field.path + display_field.field))
-            elif display_field.aggregate == "Max":
-                queryset = queryset.annotate(Max(display_field.path + display_field.field))
-            elif display_field.aggregate == "Min":
-                queryset = queryset.annotate(Min(display_field.path + display_field.field))
-            elif display_field.aggregate == "Count":
-                queryset = queryset.annotate(Count(display_field.path + display_field.field))
-            elif display_field.aggregate == "Sum":
-                queryset = queryset.annotate(Sum(display_field.path + display_field.field))
+    def add_aggregates(self, queryset, display_fields=None):
```
Comment (saulshanabrook) [inline report_builder/models.py]: Its unclear if this method mutates `self` or just returns an updated queryset. If it doesn't modify `self` than I would reccomend making this a method on a custom manager instead of a method on the model: https://docs.djangoproject.com/en/1.8/topics/db/managers/#modifying-initial-manager-querysets

File: report_builder/models.py [hunk 3]
```diff
@@ -138,6 +141,153 @@ def get_good_display_fields(self):
                 bad_display_fields.append(display_field)
         return display_fields.exclude(id__in=[o.id for o in bad_display_fields])
 
+    def report_to_list(self, queryset=None, user=None, preview=False):
```
Comment (saulshanabrook) [inline report_builder/models.py]: Same as comment above. If it just depends on the queryset and arguments, and doesn't mutate self, then make that a manager method https://docs.djangoproject.com/en/1.8/topics/db/managers/#modifying-initial-manager-querysets

Either way, the docstring should give a little better idea of what it does. What sort of list does it returns?

Comment (saulshanabrook) [inline report_builder/models.py]: Also this is a pretty long method. It would be a lot more readable if it could be broken up.

File: report_builder/models.py [hunk 4]
```diff
@@ -340,6 +494,13 @@ def choices_dict(self):
                 choices_dict.update({choice[0]: choice[1]})
         return choices_dict
 
+    @property
+    def choices(self):
```
Comment (saulshanabrook) [inline report_builder/models.py]: I think this is a reptation of property on line 451. unless i am going crazy, they have the same name

File: report_builder/models.py [hunk 5]
```diff
@@ -465,5 +626,16 @@ def filter_property(self, value):
             return not filtered
         return filtered
 
+    @property
+    def field_type(self):
+        return self.report.get_field_type(self.field, self.path)
+
+    @property
+    def choices(self):
```
Comment (saulshanabrook) [inline report_builder/models.py]: Isnt this property also the same as the above names?

File: report_builder/tests.py [hunk 6]
```diff
@@ -327,6 +332,7 @@ def test_filter_cached_property(self):
         response = self.client.get(self.generate_url)
         self.assertNotContains(response, 'lol yes')
 
+    @unittest.skip("Custom field not yet implimented")
```
Comment (saulshanabrook) [inline report_builder/tests.py]: Should this code still be here?

File: report_builder/tests.py [hunk 7]
```diff
@@ -551,6 +559,7 @@ def test_groupby_id(self):
 
         data = '"data":[[1,"John","Doe",3],[3,"Donald","King",6],[2,"Maria","Smith",2],["TOTALS","","",""],["","","",11.0]]'
 
+        print(response.data)
```
Comment (saulshanabrook) [inline report_builder/tests.py]: are the print statements in the tests still necessary?

Comment (saulshanabrook) [inline report_builder/tests.py]: if you need more debugging info i would just have a custom error message if the test failed with extra information

File: report_builder/utils.py [hunk 8]
```diff
@@ -31,3 +34,65 @@ def duplicate(obj, changes=None):
             except:
                 pass
     return duplicate
+
+
+DATE = 1
```
Comment (saulshanabrook) [inline report_builder/utils.py]: would an enum work here instead of magic numbers? https://docs.python.org/3/library/enum.html

File: report_builder/utils.py [hunk 9]
```diff
@@ -31,3 +34,65 @@ def duplicate(obj, changes=None):
...
+        if sort_type == DATE:
+            result = datetime.date(datetime.MINYEAR, 1, 1)
+        elif sort_type == NUMBER:
+            result = 0
+        else:  # Last try - make it a string
+            result = ''
+    return result
+
+
+def sort_data(data_list, display_field):
```
Comment (saulshanabrook) [inline report_builder/utils.py]: probably would make sense to add docstring describing how it is sorted.

File: report_builder/utils.py [hunk 10]
```diff
@@ -31,3 +34,65 @@ def duplicate(obj, changes=None):
...
+    elif isinstance(sample_data, (int, float, complex)):
+        sort_type = NUMBER
+    return sorted(
+        data_list,
+        key=lambda x: sort_helper(x, position, sort_type),
+        reverse=is_reverse
+    )
+
+
+def increment_total(display_field, data_row):
```
Comment (saulshanabrook) [inline report_builder/utils.py]: maybe it would make sense to move this to a [static method](http://stackoverflow.com/a/735978/907060) of a model. it seems linked very closely with model data and isnt a very generic gunction

File: report_builder/utils.py [hunk 11]
```diff
@@ -31,3 +34,65 @@ def duplicate(obj, changes=None):
...
+
+def formatter(value, style):
+    # Convert value to Decimal to apply numeric formats.
+    try:
+        value = Decimal(value)
+    except Exception:
+        pass
+
+    try:
+        return style.string.format(value)
```
Comment (saulshanabrook) [inline report_builder/utils.py]: What is this style object?

File: report_builder/models.py (hunk 2 above)
Comment (grantjenks) [inline report_builder/models.py]: Take a look at how I refactored it in django-report-utils:

https://github.com/burke-software/django-report-utils/blob/master/report_utils/mixins.py#L161

Comment (bufke) [inline report_builder/models.py]: I did review your work @grantjenks and this should look very familiar - thanks for that change.

It may be easier to read without seeing the diff.
https://github.com/burke-software/django-report-builder/blob/master/report_builder/models.py#L91

It doesn't affect self. It uses self only to gather display fields if they are not explicitly passed. Saul if this isn't clear let me know.

Comment (saulshanabrook) [inline report_builder/models.py]: Would it be possible to write this as a custom method on a manager? I think it would be clearer this way.

**Comment by grantjenks** [comment]
@bufke I tried install the RC. I hit a couple snags.

* First issue was some conflict in the migrations. Not sure how this happened. Looks like `0003_auto_20150707_0304` and `0003_auto_20150720_1549` do the same thing and conflict. They're easy enough to merge and apply.

```
  511  pip uninstall django-report-utils
  512  pip uninstall django-report-builder
  514  pip install https://pypi.python.org/packages/source/d/django-report-builder/django-report-builder-3.2.0rc2.tar.gz#md5=3519db1bc7097dc36a28e0bf515990d0
  517  python manage.py runserver --settings=crm.local

You have unapplied migrations; your app may not work properly until they are applied.
Run 'python manage.py migrate' to apply them.

  518  python manage.py migrate

CommandError: Conflicting migrations detected (0003_auto_20150707_0304, 0003_auto_20150720_1549 in report_builder).
To fix them run 'python manage.py makemigrations --merge'

  520  python manage.py makemigrations --merge

Merging report_builder
  Branch 0003_auto_20150707_0304
    - Alter field filter_type on filterfield
  Branch 0003_auto_20150720_1549
    - Alter field filter_type on filterfield

Merging will only work if the operations printed above do not conflict
with each other (working on different fields or models)
Do you want to merge these migration branches? [y/N] y

Created new merge migration /Users/grantj/repos/harvesthome/env27/lib/python2.7/site-packages/report_builder/migrations/0004_merge.py

  521  python manage.py migrate

Running migrations:
  Applying report_builder.0003_auto_20150720_1549... OK
  Applying report_builder.0004_merge... OK

  522  python manage.py runserver --settings=crm.local
```

* Second issue was a `NoReverseMatch` error.

```
NoReverseMatch at /manage/report_builder/report/
Reverse for 'report_download_xlsx' with arguments '(15,)' and keyword arguments '{}' not found. 0 pattern(s) tried: []
Request Method:	GET
Request URL:	http://127.0.0.1:8000/manage/report_builder/report/
Django Version:	1.7.7
Exception Type:	NoReverseMatch
Exception Value:	
Reverse for 'report_download_xlsx' with arguments '(15,)' and keyword arguments '{}' not found. 0 pattern(s) tried: []
```

Maybe a bad merge? Looks like there's a download-file url pattern that is meant to be used instead:

```python
    url('^report/(?P<pk>\d+)/download_file/(?P<filetype>.+)/$', views.DownloadFileView.as_view(), name="report_download_file"),
```

**Comment by bufke** [comment]
@grantjenks thanks for taking a look. I don't (see any 0003_auto_20150707_0304)[https://github.com/burke-software/django-report-builder/tree/master/report_builder/migrations] though. 

I can't reproduce the NoReverseMatch either. I believe there was a recent change that allows more than only xlsx to be used as the format. If you've customized the interface you may need to update the link to /report_builder/report/1/download_file/xlsx/ 
I made a small change to allow the url without xlsx and default to xlsx - I could see people saving these links potentially and the backwards compatibility is trivial.

File: report_builder/tests.py (hunk 6 above)
Comment (bufke) [inline report_builder/tests.py]: Yes - probably going to remove custom field support. That project never took off really and I think a hstore based solution would be better. Leaving it alone for now.

File: report_builder/models.py (hunk 5 above)
Comment (bufke) [inline report_builder/models.py]: ?
